pinned items fixes #10012

TNychka opened the pull request:
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #10012
When you pin any of the following items to your Rewards column and then buy it from that column, it unpins itself as soon as the purchase has been made:
The correct behavior is that it should remain pinned until the user chooses to unpin it.

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
Added a check on the type of item in buy to ensure that only gear items are removed.
**Assumes that ALL quests, eggs, potions, bundles, food, quests should remain pinned after purchase.**


[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: f57bd235-ec21-4357-b5c9-a0536294b285

TNychka committed:
Don't unpin non-gear items

Assumes that multiple of bundles, quests, eggs, potions can be bought

TNychka committed:
Added tests

negue commented on website/common/script/ops/buy/purchase.js:
```diff
@@ -144,8 +144,11 @@ module.exports = function purchase (user, req = {}, analytics) {
-  let itemInfo = getItemInfo(user, type, item);
-  removeItemByPath(user, itemInfo.path);
+  let singlePurchaseTypes = ['gear'];
+  if (singlePurchaseTypes.indexOf(type) !== -1) {
```
this line kinda confused me a bit 😁 , this could be converted to `.includes(type)` which is more understandable right away

TNychka replied:
There was another line of code (132-133 purchase.js) which uses a similar style, would you like me to switch that one too? It's unrelated to this issue but it's in the same function.
```
 let acceptedTypes = ['eggs', 'hatchingPotions', 'food', 'quests', 'gear', 'bundles'];
  if (acceptedTypes.indexOf(type) === -1) {
```

negue commented on website/common/script/ops/buy/purchase.js:
```diff
@@ -144,8 +144,11 @@ module.exports = function purchase (user, req = {}, analytics) {
 
-  let itemInfo = getItemInfo(user, type, item);
-  removeItemByPath(user, itemInfo.path);
+  let singlePurchaseTypes = ['gear'];
```
could you move this types-array and `acceptedTypes` outside of the method? it doesn't need to be initialized on every method call since its static anyway

negue reviewed (changes requested):
the fix looks correct, just a few requests and could also check the lint issue ?

TNychka committed:
Changed type checking and made variables global

TNychka commented:
Alrighty, I updated both my own and the previous check to use .includes now. Also I moved the variables outside the function, I'm not sure how else to prevent redeclaration. Is this the correct way to do that? Basing [this off of this](https://stackoverflow.com/questions/35271157/are-variables-initialized-inside-function-declaration-always-re-initialized-when?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa)

TNychka committed:
Lint fix

TNychka commented:
Should be good now. The coverage decreased due to moving those variables outside the function, can't be covered with tests as far as I'm aware.

negue commented:
kinda odd that this change decreases coverage

paglias commented:
thanks for this @TNychka , notated towards your next tier
